### PR TITLE
Update Android Gradle Plugin to 3.0.0B2 and Gradle 4.1

### PR DESCRIPTION
**Pull request by SalvatoreT**
All of the fun bash script changes are auto-generated using the `gradle wrapper` command.

### CODE REVIEW

**Comment by SalvatoreT** [comment]
cc @JakeWharton

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Tue Oct 20 20:22:56 EDT 2015
+#Sun Jul 02 16:40:06 PDT 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.3-all.zip
```
Comment (JakeWharton) [inline gradle/wrapper/gradle-wrapper.properties]: 4.0?

**Comment by SalvatoreT** [comment]
```

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':picasso:checkstyle'.
> Unable to create a Checker: configLocation {/home/travis/build/square/picasso/checkstyle.xml}, classpath {null}.
```

**Comment by JakeWharton** [comment]
I'll look this week

File: build.gradle [hunk 2]
```diff
@@ -2,10 +2,17 @@ subprojects {
   buildscript {
     repositories {
       mavenCentral()
+      jcenter()
+      maven {
+        url 'https://maven.google.com'
```
Comment (SimonVT) [inline build.gradle]: You can just use google().

**Comment by NightlyNexus** [comment]
Pretty sure you just have to remove the RedundantThrows in checkstyle.

**Comment by jrodbx** [comment]
See here for an example of what's needed for the checkstyle error:
https://github.com/JakeWharton/timber/pull/236/commits/87b2e9360ed4152718937e9cd9917a1d6d50c634

File: picasso/build.gradle [hunk 3]
```diff
@@ -27,21 +27,26 @@ android {
...
-  compile dep.okhttp
-  compile dep.supportAnnotations
-  testCompile dep.junit
-  testCompile dep.fest
-  testCompile dep.festAndroid
-  testCompile dep.supportV4
-  testCompile dep.robolectric
-  testCompile dep.mockito
-  testCompile dep.mockWebServer
+  implementation dep.okhttp
```
Comment (JakeWharton) [inline picasso/build.gradle]: This needs to be API

Comment (NightlyNexus) [inline picasso/build.gradle]: also the support annotations?

Comment (JakeWharton) [inline picasso/build.gradle]: Maybe. We might as well.

Comment (jrodbx) [inline picasso/build.gradle]: shouldn't support annotations be compileOnly?

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,6 +1,5 @@
-#Tue Oct 20 20:22:56 EDT 2015
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-bin.zip
```
Comment (jaredsburrows) [inline gradle/wrapper/gradle-wrapper.properties]: Can you make this `4.1-all`?

Comment (SalvatoreT) [inline gradle/wrapper/gradle-wrapper.properties]: yep yep

File: checkstyle.xml [hunk 5]
```diff
@@ -90,7 +90,7 @@
     <!-- Modifier Checks                                    -->
     <!-- See http://checkstyle.sf.net/config_modifiers.html -->
     <module name="ModifierOrder"/>
-    <module name="RedundantModifier"/>
+    <!--<module name="RedundantModifier"/>-->
```
Comment (SalvatoreT) [inline checkstyle.xml]: We either need this commented out/removed, or we need to go back and remove a lot of instances of `public`. Once you fix these 9 instances, more pop-up.
```
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/Action.java:29:5: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/AssetRequestHandler.java:37:3: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/Dispatcher.java:460:5: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/MarkableInputStream.java:39:3: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/MarkableInputStream.java:43:3: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/NetworkRequestHandler.java:34:3: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/PicassoExecutorService.java:95:5: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/Stats.java:130:5: Redundant 'public' modifier. [RedundantModifier]
[ant:checkstyle] [ERROR] /home/travis/build/square/picasso/picasso/src/main/java/com/squareup/picasso/Utils.java:348:5: Redundant 'public' modifier. [RedundantModifier]
```

**Comment by SalvatoreT** [comment]
I applied feedback from @JakeWharton, @SimonVT, @NightlyNexus, @jrodbx, and @jaredsburrows (after a month and a half) and the build is now green.

**Comment by jaredsburrows** [comment]
Awesome!

File: checkstyle.xml (hunk 5 above)
Comment (JakeWharton) [inline checkstyle.xml]: I'd like to get them fixed in a follow up then.

File: picasso-pollexor/build.gradle [hunk 6]
```diff
@@ -31,14 +31,14 @@ android {
...
-  compile project(':picasso')
-  compile dep.pollexor
-  compile dep.supportAnnotations
-  testCompile dep.junit
-  testCompile dep.robolectric
-  testCompile dep.fest
-  testCompile dep.festAndroid
-  testCompile dep.pollexor
+  implementation project(':picasso')
+  implementation dep.pollexor
```
Comment (JakeWharton) [inline picasso-pollexor/build.gradle]: Both of these should be `api` as types from both are present in the public API of this module.

**Comment by jaredsburrows** [comment]
Once this is in, I am hell with the checkstyle follow up.

**Comment by SalvatoreT** [comment]
@JakeWharton, this feels like a winner.

**Comment by JakeWharton** [comment]
Thanks for sticking with it!

**Pull request merged**